[RPC tests] Capture subprocess output and only print it on failure

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45443 [DO NOT COMMIT] Break some tests to check instrumentation
* #45442 [RPC tests] Have GDB print the stack traces of all threads upon deadlock
* #45441 [RPC tests] Log backtrace in case of SEGFAULT
* #45440 [RPC tests] Print verbose logging for TensorPipe
* **#45439 [RPC tests] Capture subprocess output and only print it on failure**

RPC tests are quite verbose (and, later in this stack, will become even more so) which makes it hard to navigate the logs to find the relevant part for a failing test. This PR redirects the stdout and stderr of child processes (i.e., which is where we run RPCs from) to some files, and only copies the content of these files to the "real" stdout if the test fails (otherwise the files are ignored). This means that successful tests won't print anything.

There were two options to do this: use a single file for all subprocesses' stdout and stderr, so that it would all get merged into a single "stream" (which is basically the effect we had before), or use separate files so that the logs of the various processes can be kept separate. The former option has the benefit of preserving the relative order of messages of different processes, while the latter option avoids ambiguity and conflicts. (In fact, many log lines are already tagged with a timestamp and a PID, so we can reconstruct some of that information after the fact). In the end, I went for the latter option, but let me know if the former is preferable.

Differential Revision: [D23963476](https://our.internmc.facebook.com/intern/diff/D23963476/)

lw committed:
[RPC tests] Capture subprocess output and only print it on failure

RPC tests are quite verbose (and, later in this stack, will become even more so) which makes it hard to navigate the logs to find the relevant part for a failing test. This PR redirects the stdout and stderr of child processes (i.e., which is where we run RPCs from) to some files, and only copies the content of these files to the "real" stdout if the test fails (otherwise the files are ignored). This means that successful tests won't print anything.

There were two options to do this: use a single file for all subprocesses' stdout and stderr, so that it would all get merged into a single "stream" (which is basically the effect we had before), or use separate files so that the logs of the various processes can be kept separate. The former option has the benefit of preserving the relative order of messages of different processes, while the latter option avoids ambiguity and conflicts. (In fact, many log lines are already tagged with a timestamp and a PID, so we can reconstruct some of that information after the fact). In the end, I went for the latter option, but let me know if the former is preferable.

Differential Revision: [D23963476](https://our.internmc.facebook.com/intern/diff/D23963476/)

[ghstack-poisoned]

lw committed:
Update on "[RPC tests] Capture subprocess output and only print it on failure"


RPC tests are quite verbose (and, later in this stack, will become even more so) which makes it hard to navigate the logs to find the relevant part for a failing test. This PR redirects the stdout and stderr of child processes (i.e., which is where we run RPCs from) to some files, and only copies the content of these files to the "real" stdout if the test fails (otherwise the files are ignored). This means that successful tests won't print anything.

There were two options to do this: use a single file for all subprocesses' stdout and stderr, so that it would all get merged into a single "stream" (which is basically the effect we had before), or use separate files so that the logs of the various processes can be kept separate. The former option has the benefit of preserving the relative order of messages of different processes, while the latter option avoids ambiguity and conflicts. (In fact, many log lines are already tagged with a timestamp and a PID, so we can reconstruct some of that information after the fact). In the end, I went for the latter option, but let me know if the former is preferable.

Differential Revision: [D23963476](https://our.internmc.facebook.com/intern/diff/D23963476/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc_utils.py:
```diff
@@ -49,8 +52,38 @@
+    def on_test_failure(self):
+        for rank in range(self.world_size):
+            try:
+                with open(os.path.join(self.output_dir, f"rank_{rank}_stdout"), "rt") as f:
```
Are these extra files cleaned up anywhere? After a test is over, they shouldn't stick around on the user's machine, right?

rohan-varma reviewed (commented):
I think the idea of the change is good, although does this mean that if I add `LOG(INFO)` to C++ RPC files while debugging then it won't be printed out if the test passes? I usually rely on LOG statements to help debug (its useful even if a test is passing).

rohan-varma commented on torch/testing/_internal/distributed/rpc_utils.py:
```diff
@@ -49,8 +52,38 @@
+        self.subprocess_init_data["output_dir"] = self.output_dir
+
+    def subprocess_init(self, data):
+        with open(os.path.join(data["output_dir"], f"rank_{self.rank}_stdout"), "wt") as f:
```
In general, it looks like we redirect stdout and err to different files, and then undo that redirection on a failure to print out the errors. Would a context manager similar to the one the one defined here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/distributed_test.py#L136 work for this use case? It might help reduce some of the file handling logic.

lw committed:
Update on "[RPC tests] Capture subprocess output and only print it on failure"


RPC tests are quite verbose (and, later in this stack, will become even more so) which makes it hard to navigate the logs to find the relevant part for a failing test. This PR redirects the stdout and stderr of child processes (i.e., which is where we run RPCs from) to some files, and only copies the content of these files to the "real" stdout if the test fails (otherwise the files are ignored). This means that successful tests won't print anything.

There were two options to do this: use a single file for all subprocesses' stdout and stderr, so that it would all get merged into a single "stream" (which is basically the effect we had before), or use separate files so that the logs of the various processes can be kept separate. The former option has the benefit of preserving the relative order of messages of different processes, while the latter option avoids ambiguity and conflicts. (In fact, many log lines are already tagged with a timestamp and a PID, so we can reconstruct some of that information after the fact). In the end, I went for the latter option, but let me know if the former is preferable.

Differential Revision: [D23963476](https://our.internmc.facebook.com/intern/diff/D23963476/)

[ghstack-poisoned]

lw committed:
Update on "[RPC tests] Capture subprocess output and only print it on failure"


RPC tests are quite verbose (and, later in this stack, will become even more so) which makes it hard to navigate the logs to find the relevant part for a failing test. This PR redirects the stdout and stderr of child processes (i.e., which is where we run RPCs from) to some files, and only copies the content of these files to the "real" stdout if the test fails (otherwise the files are ignored). This means that successful tests won't print anything.

There were two options to do this: use a single file for all subprocesses' stdout and stderr, so that it would all get merged into a single "stream" (which is basically the effect we had before), or use separate files so that the logs of the various processes can be kept separate. The former option has the benefit of preserving the relative order of messages of different processes, while the latter option avoids ambiguity and conflicts. (In fact, many log lines are already tagged with a timestamp and a PID, so we can reconstruct some of that information after the fact). In the end, I went for the latter option, but let me know if the former is preferable.

Differential Revision: [D23963476](https://our.internmc.facebook.com/intern/diff/D23963476/)

[ghstack-poisoned]

pritamdamania87 commented:
I'm wondering if it makes sense to prepend each log line with the rank/process id to make parsing through the logs easier? I'm assuming in the successful case it doesn't really matter how much you print and in the failure case we're anyways copying everything to stdout.

If that is the case, why don't we have all the verbose logging on stderr and stdout and developers can always pipe this to a file and grep through if needed. Or is there a different use case here that I am missing?
